whats-that-typeclass-Foldable/article.md

```diff
@@ -0,0 +1,364 @@
+
+```haskell
+instance Foldable Maybe where
+  foldr :: (a -> b -> b) -> b -> Maybe a -> b
+  foldr _ z Nothing  = z
+  foldr f z (Just x) = f x z
+```
+
+Usage examples:
+
+```haskell
+ghci> foldr (+) 1 (Just 2)
+3
+ghci> foldr (+) 1 Nothing
+1
+```
+
+#### `BinarySearchTree a`
+
+A more interesting and useable `Foldable` instance could be defined for a `BinarySearchTree a` data type. Consider [a binary search tree](https://en.wikipedia.org/w/index.php?title=Binary_search_tree&oldid=1053783914), each node of which either stores a value of type `a` and has left and right subtrees, or is just a leaf without a value. Moreover, for each non-leaf node `n`, all values in its left subtree should be less or equal than the `n`'s value, and all values in its right subtree should be greater than the `n`'s value. 
+
+![Binary search tree for the integer numbers](./bst.png)
+
+This structure matches the following Haskell definition:
+
+```haskell
+data BinarySearchTree a
+  = Branch (BinarySearchTree a) a (BinarySearchTree a)
+  | Leaf
+```
+
+Imagine that we want to reduce the whole tree to just one value. We could sum values of nodes, multiply them, or perform any other binary operation. So,it's reasonable to define a folding function that matches your goals. We suggest you try to implement `Foldable` instance on your own; it's one of the exercises below. Take into account that to define `foldr`, we need to go through the tree's elements from right to left — from right subtree to left subtree through the non-leaf node connecting them.
+
+## Minimal Complete Definition of `Foldable` typeclass
+
+### `foldMap`
+
+The next important method of `Foldable` is `foldMap`. It has the following declaration:
+
+```haskell
+foldMap :: (Monoid m, Foldable t) => (a -> m) -> t a -> m
+```
+
+It's important to note that `foldMap` has a [`Monoid`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#t:Monoid) constraint. Its first argument is a function that maps each element of a container into a monoid, the second argument is a container itself. After mapping elements, the results are combined using `(<>)` operator. The order of folding is from right to left, so `foldMap` could be implemented via `foldr`. 
+
+Let's look at how exactly that could be done:
+
+```haskell
+foldMap :: (Monoid m, Foldable t) => (a -> m) -> t a -> m
+foldMap f = foldr (\x acc -> f x <> acc) mempty
+```
+
+`foldMap` doesn't have a base element, as only the elements of a container are reduced. However, `foldr` does, so it perfectly makes sense to use the identity of monoid — `mempty`. We can also see that the folding function `f` is composed with `(<>)`, thus, the current result is appended to a monoid accumulator on each step.
+
```

Comment:
I would also mention some example of such a `Monoid`, e.g. `Sum`, and briefly illustrate how `foldMap` on it turns out to do exactly the same as the respective `foldr`.

Reply:
[Done](https://github.com/serokell/blog-posts/pull/3/commits/34d47e614d713cb423154df8aefce49507ead369)

Reply:
Yeah. I would mention, that `foldMap` and `foldr` are interchangeable, it is just that the former picks reducer and zero point from `Monoid` instance, and the latter accepts them explicitly - or something like that (phrasing is subject to change).

Reply:
And so you can show that `foldMap Sum` is the same as `foldr (+) 0`.

Reply:
> And so you can show that foldMap Sum is the same as foldr (+) 0.

Oh, I see that you meant exactly this when wrote:

> e.g. Sum, and briefly illustrate how foldMap on it turns out to do exactly the same as the respective foldr.

Now it seems a bit odd to have these both examples, like one is enough, and `foldMap Sum == foldr (+) 0` looks quite easier than summing squares :/

Reply:
I've come up with idea! We can leave `foldMap Sum == foldr (+) 0`  as an easy example here, and move summing squares to the Exercise part (show implementation using `foldr` and ask to implement via `foldMap` or vice versa).

How does it sound? 🙃

Reply:
Yeah, sounds awesome.

Reply:
[Pushed](https://github.com/serokell/blog-posts/pull/3/commits/12f3d8dd6f08dc00745bd5f01619e4f99b9210cf). Hope the exercise's solutions are not weird 😅

Reply:
Thanks, looks very good to me :+1: